com.unity.render-pipelines.universal/Runtime/ForwardLights.cs

```diff
@@ -75,6 +98,201 @@ public ForwardLights()
                 m_AdditionalLightOcclusionProbeChannels = new Vector4[maxLights];
                 m_AdditionalLightsLayerMasks = new float[maxLights];
             }
+
+            if (m_UseClusteredRendering)
+            {
+                m_ZBinBuffer = new ComputeBuffer(UniversalRenderPipeline.maxZBins / 4, UnsafeUtility.SizeOf<float4>(), ComputeBufferType.Constant, ComputeBufferMode.Dynamic);
+                m_TileBuffer = new ComputeBuffer(UniversalRenderPipeline.maxTileVec4s, UnsafeUtility.SizeOf<float4>(), ComputeBufferType.Constant, ComputeBufferMode.Dynamic);
+                m_RequestedTileWidth = tileSize;
+            }
+        }
+
+        internal void ProcessLights(ref RenderingData renderingData)
+        {
+            if (m_UseClusteredRendering)
+            {
+                var camera = renderingData.cameraData.camera;
+                var screenResolution = math.int2(renderingData.cameraData.pixelWidth, renderingData.cameraData.pixelHeight);
+
+                var lightCount = renderingData.lightData.visibleLights.Length;
+                var lightOffset = 0;
+                while (lightOffset < lightCount && renderingData.lightData.visibleLights[lightOffset].lightType == LightType.Directional)
+                {
+                    lightOffset++;
+                }
+                if (lightOffset == lightCount) lightOffset = 0;
+                lightCount -= lightOffset;
+
+                m_DirectionalLightCount = lightOffset;
+                if (renderingData.lightData.mainLightIndex != -1) m_DirectionalLightCount -= 1;
+
+                var visibleLights = renderingData.lightData.visibleLights.GetSubArray(lightOffset, lightCount);
+                var lightsPerTile = UniversalRenderPipeline.lightsPerTile;
+                var wordsPerTile = lightsPerTile / 32;
+
+                m_ActualTileWidth = m_RequestedTileWidth >> 1;
+                do
+                {
+                    m_ActualTileWidth = m_ActualTileWidth << 1;
+                    m_TileResolution = (screenResolution + m_ActualTileWidth - 1) / m_ActualTileWidth;
```

Comment:
That wouldn't work because `screenResolution / m_ActualTileWidth` would round down the result, so there would be nothing to ceil. We can add a utility method for integer division that rounds up though.